### PR TITLE
Hydrophilic quirk, Species whitelists for quirks!

### PR DESCRIPTION
**Pull request by mcmeiler**
## About The Pull Request

Adds a new trait called Hydrophilic, whitelisted to slime hybrids.

The quirk causes you to lose blood whenever you're hit by water.

Adds functional quirk whitelists for species. For an example, check `code/controllers/subsystem/processing/quirks.dm`.

## Screenshots and Testing Procedure

Selecting the quirk as a human:
image

Selecting the quirk as a slimeperson:
image
image

## Changelog 

:cl: mc_meiler, suggestion by FlamesWillLive
add: New slime-only quirk: Hydrophilic. You melt in water. Good luck.
code: Added whitelisting functionality for quirks.
/:cl:

### CODE REVIEW

**Comment by mcmeiler** [comment]
Holy _shit_. I can _go to sleep_. GOOD NIGHT BUBBERSTATION.

**Comment by mcmeiler** [comment]
Pretty much ready for TM at this point.

**Comment by mcmeiler** [comment]
_Now_ it's pretty much ready for TM. I've ensured everything works on local - whitelists and the quirk itself.

File: modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/datum/quirk/hydrophilic
+	name = "Hydrophilic"
+	desc = "(Slimeperson only) Your molecules are highly hydrophilic - or, in layman's terms, dissolve very well in water."
+	icon = "bug" // I can't be asked to make an icon for this, my spritework sucks.
+	value = 0
+	quirk_flags = QUIRK_HUMAN_ONLY
```
Comment (Cyprex) [inline modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm:6]: ?

File: code/datums/quirks/_quirk.dm [hunk 2]
```diff
@@ -67,6 +69,12 @@
 	if(quirk_holder)
 		CRASH("Attempted to add quirk to a holder when it already has a holder.")
 
+	//BUBBER EDIT ADDITION START - Species quirks
+	//We silently fail if a whitelisted quirk *somehow* got into someone's preferences.
+	if(quirk_whitelist_flags & QUIRK_SLIMEPERSON_ONLY && !isroundstartslime(new_holder))
+		return
```
Comment (Cyprex) [inline code/datums/quirks/_quirk.dm:75]:
```suggestion
		CRASH("Attempted to add quirk [name] to [quirk_holder], but the species is ineligible for it!")
```

Comment (Cyprex) [reply code/datums/quirks/_quirk.dm:75]: I'm fairly sure this just runtimes, i. e. returns, and we ought to fail loudly.

Comment (Cyprex) [reply code/datums/quirks/_quirk.dm:75]: https://www.byond.com/docs/ref/#/proc/CRASH

File: modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm (hunk 1 above)
Comment (Cyprex) [reply modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm:6]: Oh, I see, this is for the type, not the species. I suppose this prevents it from applying to.... borgs? 🤔

File: modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/datum/quirk/hydrophilic
+	name = "Hydrophilic"
+	desc = "(Slimeperson only) Your molecules are highly hydrophilic - or, in layman's terms, dissolve very well in water."
```
Comment (Cyprex) [inline modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm:3]:
```suggestion
	desc = "(Slimeperson only) Your molecules are highly hydrophilic - or, in layman's terms, dissolve very well in water. You should probably stay away from it..."
```

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx [hunk 4]
```diff
@@ -201,6 +203,29 @@ export const QuirksPage = (props, context) => {
...
+            const whitelistSpeciesName = whitelist[1];
+            // Quirk is not in the whitelisted ones, continue to next list
+            if (whitelist.indexOf(quirk.name) === -1) {
+              continue;
+            }
+
+            // If the selected species is NOT in the list, do not give the quirk
+            // Otherwise, give the quirk
+            if (whitelist.indexOf(currentSpeciesName) === -1) {
+              return `This quirk is whitelisted to ${whitelistSpeciesName}!`;
```
Comment (Cyprex) [inline tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:223]:
```suggestion
              return `This quirk can only be taken by ${whitelistSpeciesName}!`;
```

Comment (Cyprex) [reply tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:223]: or is only available to, something along those lines.

**Review by Cyprex — COMMENTED** [review]
I hate the JS side of this but the blacklist is implemented in the same manner & I don't want to pile even more onto you, so whatever.

File: modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm (hunk 1 above)
Comment (mcmeiler) [reply modular_zubbers/code/datums/bubber_quirks/hydrophilic.dm:6]: Borgs, simplemobs, etc.

**Comment by mcmeiler** [comment]
> I hate the JS side of this but the blacklist is implemented in the same manner

I also hate the JS side of this and I'll look into making it better with a followup PR, probably gonna fix the awful formatting of the whitelist list as well.

**Comment by Cyprex** [comment]
If so upstream code should be edited upstream, and *maybe* we could even get the species whitelist there, dunno.
